presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -246,19 +266,51 @@ private Estimate calculateNullsFraction(Map<String, PartitionStatistics> statist
                     if (partitionsWithStatisticsCount == 0) {
                         return OptionalDouble.empty();
                     }
-                    else {
-                        return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
-                    }
+                    return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
                 });
 
-        if (totalNullsCount.isValueUnknown() || totalRowsCount.isValueUnknown()) {
+        if (totalNullsCount.isValueUnknown()) {
             return Estimate.unknownValue();
         }
-        if (totalRowsCount.getValue() == 0.0) {
-            return Estimate.zeroValue();
+        return new Estimate(totalNullsCount.getValue() / rowCount.getValue());
+    }
+
+    private Estimate calculateDataSize(
+            Map<String, PartitionStatistics> statisticsSample,
+            String columnName,
+            Estimate rowCount,
+            Estimate nullsFraction,
+            OptionalDouble rowsPerPartition)
+    {
+        if (rowCount.isValueUnknown() || !rowsPerPartition.isPresent()) {
+            return Estimate.unknownValue();
+        }
+
+        double knownNonNullRowCount = 0;
+        double knownDataSize = 0;
+
+        for (PartitionStatistics statistics : statisticsSample.values()) {
+            double partitionRowCount = orElse(statistics.getBasicStatistics().getRowCount(), rowsPerPartition.getAsDouble());
+            HiveColumnStatistics partitionColumnStatistics = statistics.getColumnStatistics().get(columnName);
+            if (partitionColumnStatistics == null || !partitionColumnStatistics.getAverageColumnLength().isPresent()) {
+                continue;
+            }
+            double partitionNonNullCount = partitionRowCount - partitionColumnStatistics.getNullsCount().orElse(0);
+            if (partitionNonNullCount < 0) {
+                log.debug("null count bigger than row count in partition stats");
+                continue;
+            }
+            knownNonNullRowCount += partitionNonNullCount;
+            // Note: average column length from Hive might not translate directly into internal data size
+            knownDataSize += partitionColumnStatistics.getAverageColumnLength().getAsDouble() * partitionNonNullCount;
+        }
+
+        if (knownNonNullRowCount == 0) {
+            return Estimate.unknownValue();
         }
 
-        return new Estimate(totalNullsCount.getValue() / totalRowsCount.getValue());
+        double nonNullCount = rowCount.getValue() * (1 - (nullsFraction.isValueUnknown() ? 0 : nullsFraction.getValue()));
+        return new Estimate(knownDataSize / knownNonNullRowCount * nonNullCount);
```

Comment:
Is is so much more complex than: https://github.com/starburstdata/presto-private/blob/0a47179b9d0df73724863fba226372a31c9355f5/presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java#L381 ?
Why it is the case? Are there any practical examples where code here is better/more accurate?

Is there an advantage in computing `knownNonNullRowCount` and `knownDataSize` 
vs using `totalRowsCount` and `nullsFraction` estimates?
Keep in mind that we are still using `rowCount` and `nullsFraction` at the end so we still risk measurements errors.

Reply:
I see that we want to do weighted average from each partition. This still could be done more with in a streaming fashion.

Reply:
This method was simplified as part of the https://github.com/prestodb/presto/pull/11185. Instead of having `avgColumnLength` in the `HiveColumnStatistics` we have `totalSizeInBytes`. We convert `avgColumnLength` to `totalSizeInBytes` when pulling the statistics from the Hive metastore.